File: TreeKalk/heap.cpp

#include "heap.h"

heap::heap():AlberoBinario(){}
heap::heap(const heap& b,QString q):AlberoBinario(b.getRadice()->copy(),q){}
heap::heap(nodo* n,const QString& q):AlberoBinario(n,q){}

heap::~heap()
{
    //cout << "~alberobinario()\n";
}

bool heap::operator ==(const heap& b)const
{
    return AlberoBinario::operator ==(b);
}
bool heap::operator !=(const heap& b)const
{
    return !(operator ==(b));
}
heap& heap::operator =(const heap& b)
{
    if(this!=&b)
    {
        delete getRadice();
        setRoot(b.getRadice()->copy());
    }
    return *this;
}
QString heap::converti(int i) const
{
    QString s=QString::number(i);
    return s.setNum(i,2);
}
binary* heap::ultimo()const
{
    QString b=converti(dimensione(getRadice()));
    binary* temp=getRadice();
    int j=1;
    while(j<b.size())
    {
        if( b.mid(j,1)=="0")
            temp=temp->left();
        else if(b.mid(j,1)=="1")
            temp=temp->right();
        j++;
    }
    return temp;
}

binary* heap::trovaPrimaRicorrenza(double i)const
{
    return trovaPrimaRicorrenzaHelp(i,getRadice());
}
binary* heap::trovaPrimaRicorrenzaHelp(double i,binary* n)const
{
    if(!n)
        return NULL;
    binary* temp=NULL;
    if(i==n->getInfo())
        return n;
    if(!temp && n->left())
        temp=trovaPrimaRicorrenzaHelp(i,n->left());
    if(!temp && n->right())
        temp=trovaPrimaRicorrenzaHelp(i,n->right());
    return temp;
}

File: java/heap.java



public abstract class heap extends BST
{
    public heap()
    {
        super();
    }
    public heap(heap h)
    {
        super(h.getRadice().copy());
    }
    public heap(nodo n)
    {
        super((binary)n);
    }
    public String converti(int i)
    {
        int a;
        String x="";
        while(i > 0)
        {
            a = i % 2;
            x = a+x;
            i = i / 2;
        }
        return x;
    }
    public binary ultimo()
    {
        String b=converti(dimensione(getRadice()));
        binary temp=getRadice();
        int j=1;
        while(j<b.length())
        {
            if( b.charAt(j)=='0')
                temp=temp.left();
            else if(b.charAt(j)=='1')
                temp=temp.right();
            j++;
        }
        return temp;
    }

    public binary trovaPrimaRicorrenza(double i)
    {
        return trovaPrimaRicorrenzaHelp(i,getRadice());
    }
    public binary trovaPrimaRicorrenzaHelp(double i,binary n)
    {
        if(n==null)
            return null;
        binary temp=null;
        if(i==n.getInfo())
            return n;
        if(temp==null && n.left()!=null)
            temp=trovaPrimaRicorrenzaHelp(i,n.left());
        if(temp!=null && n.right()==null)
            temp=trovaPrimaRicorrenzaHelp(i,n.right());
        return temp;
    }
    public abstract void heapify(binary b);
    public abstract void heapifyUp(binary b);


    public static void main(String []args)
    {
    }

}
